File: hard/0023_merge_k_sorted_lists/solution.py
import heapq
from typing import List, Optional

from utils.python.nodes import ListNode


class Solution:
    def mergeKLists(
        self, lists: List[Optional[ListNode]]
    ) -> Optional[ListNode]:
        heap = []

        for head in lists:
            if head:
                heapq.heappush(heap, (head.val, head))

        dummy = ListNode(-1)
        curr = dummy

        while heap:
            _, node = heapq.heappop(heap)

            curr.next = node
            curr = curr.next

            if node.next:
                heapq.heappush(heap, (node.next.val, node.next))

        return dummy.next

File: hard/0023_merge_k_sorted_lists/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <queue>
#include <vector>

class Solution {
public:
  utils::ListNode *mergeKLists(std::vector<utils::ListNode *> &lists) {
    auto cmp = [](utils::ListNode *a, utils::ListNode *b) {
      return a->val > b->val;
    };
    std::priority_queue<utils::ListNode *, std::vector<utils::ListNode *>,
                        decltype(cmp)>
        pq(cmp);

    for (utils::ListNode *list : lists) {
      if (list)
        pq.push(list);
    }

    utils::ListNode *dummy = new utils::ListNode(-10001);
    utils::ListNode *curr = dummy;

    while (!pq.empty()) {
      curr->next = pq.top();
      pq.pop();

      curr = curr->next;
      if (curr->next)
        pq.push(curr->next);
    }
    curr->next = nullptr;

    return dummy->next;
  }
};
